File: src/python/algorithms/graph/numberOfIslands2/number_of_islands2.py
class UnionFind:
    def __init__(self, size):
        self.parent = [-1] * size
        self.rank = [0] * size

    def add_land(self, x):
        self.parent[x] = x

    def is_land(self, x):
        return self.parent[x] != -1

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return False
        if self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        elif self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1
        return True


# TC: O(k * α(m*n)) ≈ O(k),  SC: O(m*n)
def number_of_islands2(m, n, positions):
    result = []
    uf = UnionFind(m * n)
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
    islands_count = 0

    for row, col in positions:
        index = row * n + col

        if uf.is_land(index):
            result.append(islands_count)
            continue

        uf.add_land(index)
        islands_count += 1

        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if nr < 0 or nr >= m or nc < 0 or nc >= n:
                continue
            neighbour_index = nr * n + nc
            if uf.is_land(neighbour_index):
                if uf.union(index, neighbour_index):
                    islands_count -= 1

        result.append(islands_count)
    return result


test_cases = [
    (3, 3, [[0,0],[0,1],[1,2],[2,1]], [1,1,2,3]),
    (2, 2, [[0,0],[0,1],[1,0],[1,1]], [1,1,1,1]),
    (1, 1, [[0,0]], [1]),
    (3, 3, [[0,0],[0,0],[0,1]], [1,1,1]),
    (3, 3, [[0,0],[2,2],[1,1],[0,1],[1,0]], [1,2,3,2,2]),
]

for i, (m, n, positions, expected) in enumerate(test_cases, 1):
    result = number_of_islands2(m, n, positions)
    print(f"Test {i}: Result={result}, Expected={expected}")

File: src/javascript/algorithms/graph/numberOfIslands2/numberOfIslands2.js
class UnionFind {
    constructor(size) {
        this.parent = new Array(size).fill(-1);
        this.rank = new Array(size).fill(0);
    }

    addLand(x) {
        this.parent[x] = x;
    }

    isLand(x) {
        return this.parent[x] !== -1;
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX === rootY) return false;

        if (this.rank[rootX] > this.rank[rootY]) {
            this.parent[rootY] = rootX;
        } else if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else {
            this.parent[rootY] = rootX;
            this.rank[rootX]++;
        }
        return true;
    }
}

// TC: O(k * α(m*n)) ≈ O(k),  SC: O(m*n)
function numberOfIslands2(m, n, positions) {
    const result = [];
    const uf = new UnionFind(m * n);
    const directions = [[-1, 0], [0, 1], [1, 0], [0, -1]];
    let islandsCount = 0;

    for (const [row, col] of positions) {
        const index = row * n + col;

        if (uf.isLand(index)) {
            result.push(islandsCount);
            continue;
        }

        uf.addLand(index);
        islandsCount++;

        for (const [dr, dc] of directions) {
            const nr = row + dr;
            const nc = col + dc;
            if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
            const neighbourIndex = nr * n + nc;
            if (uf.isLand(neighbourIndex)) {
                if (uf.union(index, neighbourIndex)) {
                    islandsCount--;
                }
            }
        }
        result.push(islandsCount);
    }
    return result;
}

const testCases = [
    { m: 3, n: 3, positions: [[0,0],[0,1],[1,2],[2,1]], expected: [1,1,2,3] },
    { m: 2, n: 2, positions: [[0,0],[0,1],[1,0],[1,1]], expected: [1,1,1,1] },
    { m: 1, n: 1, positions: [[0,0]], expected: [1] },
    { m: 3, n: 3, positions: [[0,0],[0,0],[0,1]], expected: [1,1,1] },
    { m: 3, n: 3, positions: [[0,0],[2,2],[1,1],[0,1],[1,0]], expected: [1,2,3,2,2] },
];

testCases.forEach(({ m, n, positions, expected }, i) => {
    const result = numberOfIslands2(m, n, positions);
    console.log(`Test ${i + 1}: Result=${JSON.stringify(result)}, Expected=${JSON.stringify(expected)}`);
});
